### PR TITLE
SparseML integration - prune and quantize models during training

### PR DESCRIPTION
**Pull request by bfineran**

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -233,14 +239,41 @@ def train(hyp, opt, device, tb_writer=None):
     model.class_weights = labels_to_class_weights(dataset.labels, nc).to(device) * nc  # attach class weights
     model.names = names
 
+    qat = False
```
Comment (mgoin) [inline train.py]: nit: move this variable to right before you set it at the bottom of this block at 
```if manager.quantization_modifiers:```

File: train.py [hunk 2]
```diff
@@ -233,14 +239,41 @@ def train(hyp, opt, device, tb_writer=None):
...
+    # SparseML Integration
+    if not opt.use_leaky_relu:  # use LeakyReLU activations
```
Comment (mgoin) [inline train.py:243]: do we require leaky_relu? should we log/warn?

File: train.py [hunk 3]
```diff
@@ -233,14 +239,41 @@ def train(hyp, opt, device, tb_writer=None):
...
+        loggers=[PythonLogger(), TensorBoardLogger(writer=tb_writer)]
+    )
+    # override lr scheduler if recipe makes any LR updates
+    if manager.learning_rate_modifiers:
+        logger.info('Disabling yolo LR scheduler, managing LR using SparseML recipe')
+        scheduler = None
+    # override num epochs if recipe explicitly modifies epoch range
+    if manager.epoch_modifiers and manager.max_epochs:
+        epochs = manager.max_epochs or epochs  # override num_epochs
+        logger.info(f'overriding number of epochs from SparseML manager to {manager.max_epochs}')
```
Comment (mgoin) [inline train.py]: nit: captial "O" overriding to match other logs

File: train.py [hunk 4]
```diff
@@ -389,12 +426,15 @@ def train(hyp, opt, device, tb_writer=None):
...
+                ckpt_model = deepcopy(model.module if is_parallel(model) else model)
+                if qat:
+                    ckpt_model = model.state_dict()  # pickled QAT exports not currently supported
                 ckpt = {'epoch': epoch,
                         'best_fitness': best_fitness,
                         'training_results': results_file.read_text(),
-                        'model': deepcopy(model.module if is_parallel(model) else model).half(),
-                        'ema': deepcopy(ema.ema).half(),
-                        'updates': ema.updates,
+                        'model': ckpt_model.half() if not opt.disable_amp else ckpt_model,
```
Comment (mgoin) [inline train.py:441]: what happened to the deepcopy?

File: train.py [hunk 5]
```diff
@@ -422,23 +462,36 @@ def train(hyp, opt, device, tb_writer=None):
...
+        # ONNX export
+        if opt.export_onnx:
+            onnx_path = f'{save_dir}/model.onnx'
+            logger.info(f'training complete, exporting ONNX to {onnx_path}')
+            export_model = model.module if is_parallel_model(model) else model
+            export_model.model[-1].export = True  # do not export grid post-procesing
+            exporter = ModuleExporter(export_model, save_dir)
+            exporter.export_onnx(torch.randn(1, 3, imgsz, imgsz), convert_qat=True)
+            if qat:
+                skip_onnx_input_quantize(onnx_path, onnx_path)
```
Comment (mgoin) [inline train.py]: what is `skip_onnx_input_quantize` doing and why does it have the same `onnx_path` arg twice?

File: train.py [hunk 6]
```diff
@@ -489,6 +542,11 @@ def train(hyp, opt, device, tb_writer=None):
     parser.add_argument('--bbox_interval', type=int, default=-1, help='Set bounding-box image logging interval for W&B')
     parser.add_argument('--save_period', type=int, default=-1, help='Log model after every "save_period" epoch')
     parser.add_argument('--artifact_alias', type=str, default="latest", help='version of dataset artifact to be used')
+    parser.add_argument('--sparseml-recipe', type=str, default=None, help='Path to a SparseML sparsification recipe, see <TODO: Link Here> for more information')
+    parser.add_argument('--use-leaky-relu', action='store_true', help='Override default SiLU activation with LeakyReLU')
+    parser.add_argument('--export-onnx', action='store_true', help='export final model to ONNX')
+    parser.add_argument('--disable-amp', action='store_true', help='Disable FP16 half precision (enabled by default)')
+    parser.add_argument('--disable-ema', action='store_true', help='Disable EMA model updates (enabled by default)')
```
Comment (mgoin) [inline train.py:558]: I assume we need to disable these to get our stuff to work, can we make that more explicit or will it be in documentation?

**Review by mgoin — COMMENTED** [review]
awesome job @bfineran it is really nice to achieve all this in net ~60 lines of code!

File: train.py (hunk 1 above)
Comment (bfineran) [reply train.py]: good catch, actually meant to wrap the next section on an if statement conditional on a recipe being provided and this was supposed to be initialized outside of it.

File: train.py (hunk 2 above)
Comment (bfineran) [reply train.py:243]: It's the best tradeoff for us between speed and accuracy for pruned_quant.  The default SiLU will not export to ONNX, I'll add a warning when we detect QAT present, and also wrap the ONNX export in a try/except

File: train.py (hunk 4 above)
Comment (bfineran) [reply train.py:441]: that's now handled by the `ckpt_model` variable

File: train.py (hunk 5 above)
Comment (bfineran) [reply train.py]: this is the optimization pass to remove the initial `QuantizeLinear` node in an ONNX graph so the model can take uint8 inputs.  the first two arguments are the model file and the save file.  gave the same file for both so it will override the original.

File: train.py (hunk 6 above)
Comment (bfineran) [reply train.py:558]: plan for now is documentation, would like to offer these as generic utilities that also happen to help with our integration.  Lmk if you have thoughts on something short to add though

**Review by bfineran — COMMENTED** [review]
Thanks for the review @mgoin. Just put up some adjustments and responses

File: requirements.txt [hunk 7]
```diff
@@ -27,3 +27,4 @@ pandas
 # extras --------------------------------------
 thop  # FLOPS computation
 pycocotools>=2.0  # COCO mAP
+sparseml>=0.2
```
Comment (markurtz) [inline requirements.txt]: we may want to change this to ~ for now because we have some changes coming in that will be breaking
